validator/src/main/java/com/google/udmi/util/messaging/MessagingClient.java

```diff
@@ -0,0 +1,51 @@
+package com.google.udmi.util.messaging;
+
+import com.google.pubsub.v1.PubsubMessage;
+import java.io.Closeable;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * An interface for a generic messaging client, abstracting away the specifics
+ * of the underlying technology (e.g., Google Pub/Sub, MQTT).
+ */
+public interface MessagingClient extends Closeable {
+
+  String MQTT = "mqtt";
+  String PUBSUB = "pubsub";
+
+  /**
+   * Get appropriate messaging client from supplied config.
+   */
+  static MessagingClient from(MessagingClientConfig config) {
+    return switch (config.protocol()) {
+      case MQTT -> MqttMessagingClient.from(config);
+      case PUBSUB -> GenericPubSubClient.from(config);
+      default -> throw new UnsupportedOperationException(
+          "Unsupported messaging protocol " + config.protocol());
+    };
+  }
+
+  /**
+   * Polls for a message from the subscription, waiting up to the specified timeout.
+   *
+   * @param timeout The maximum time to wait.
+   * @param unit The time unit of the timeout argument.
+   * @return The received PubsubMessage, or null if the timeout is reached.
+   */
+  PubsubMessage poll(long timeout, TimeUnit unit);
```

Comment:
prefer `Duration` over two args -- if that doesn't mess other things up too much.